File: packages/runtime-java/src/test/java/besdk/runtime/ValidateTest.java
package besdk.runtime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class ValidateTest {

  private static List<String> check(Object value, String descriptor) {
    return Validate.check(value, Schema.of(descriptor), Map.of());
  }

  @Test
  void acceptsMatchingPrimitives() {
    assertTrue(check("x", "{\"k\":\"str\"}").isEmpty());
    assertTrue(check(1L, "{\"k\":\"int\"}").isEmpty());
    assertTrue(check(1.5, "{\"k\":\"num\"}").isEmpty());
    assertTrue(check(Boolean.TRUE, "{\"k\":\"bool\"}").isEmpty());
    assertTrue(check(null, "{\"k\":\"null\",\"i\":{\"k\":\"str\"}}").isEmpty());
  }

  @Test
  void namesTheFieldAndBothTypes() {
    List<String> problems =
        check(Map.of("id", 42L), "{\"k\":\"obj\",\"f\":[[\"id\",{\"k\":\"str\"},1]]}");
    assertEquals(List.of("id should be a string but was an integer"), problems);
  }

  @Test
  void acceptsAWholeDoubleWhereAnIntegerIsDeclared() {
    // A JSON integer arrives as 1.0 from serialisers with no integer type. Rejecting that would
    // fail on data
    // that is actually correct.
    assertTrue(check(1.0, "{\"k\":\"int\"}").isEmpty());
    assertFalse(check(1.5, "{\"k\":\"int\"}").isEmpty());
  }

  @Test
  void reportsAMissingRequiredFieldButNotAMissingOptionalOne() {
    String schema = "{\"k\":\"obj\",\"f\":[[\"id\",{\"k\":\"str\"},1],[\"name\",{\"k\":\"str\"}]]}";
    assertEquals(List.of("id is missing"), check(Map.of("name", "x"), schema));
    assertTrue(check(Map.of("id", "x"), schema).isEmpty());
  }

  @Test
  void neverComplainsAboutUnknownFields() {
    // A server adding a field must not break a client. That is the whole point of an evolving API.
    assertTrue(
        check(
                Map.of("id", "x", "brandNew", 123L),
                "{\"k\":\"obj\",\"f\":[[\"id\",{\"k\":\"str\"},1]]}")
            .isEmpty());
  }

  @Test
  void treatsAnEmptyArrayAsAValidEmptyMap() {
    // The PHP empty-map artifact: a PHP backend serialises `{}` as `[]`, and that is a valid empty
    // map rather
    // than a wrong type.
    assertTrue(check(List.of(), "{\"k\":\"map\",\"v\":{\"k\":\"str\"}}").isEmpty());
    assertTrue(check(List.of(), "{\"k\":\"obj\",\"f\":[]}").isEmpty());
  }

  @Test
  void walksArraysAndReportsTheIndex() {
    List<String> problems = check(List.of("a", 2L), "{\"k\":\"arr\",\"i\":{\"k\":\"str\"}}");
    assertEquals(List.of("[1] should be a string but was an integer"), problems);
  }

  @Test
  void acceptsAnyBranchOfAUnionAndReportsOnceWhenNoneMatch() {
    String schema = "{\"k\":\"or\",\"o\":[{\"k\":\"str\"},{\"k\":\"int\"}]}";
    assertTrue(check("x", schema).isEmpty());
    assertTrue(check(3L, schema).isEmpty());
    // One message, not one per branch: a union of five reporting five problems buries the real one.
    assertEquals(1, check(Boolean.TRUE, schema).size());
  }

  @Test
  void terminatesOnASelfReferentialSchema() {
    // The cycle closes through the table rather than through recursion, so this is finite by
    // construction
    // rather than by a depth cap.
    Map<String, Schema> table =
        Schema.table(
            "{\"Node\":{\"k\":\"obj\",\"f\":[[\"child\",{\"k\":\"ref\",\"n\":\"Node\"}]]}}");
    Object value = Map.of("child", Map.of("child", Map.of("child", Map.of())));
    assertTrue(Validate.check(value, Schema.of("{\"k\":\"ref\",\"n\":\"Node\"}"), table).isEmpty());
  }

  @Test
  void treatsAMissingTableEntryAsAny() {
    // An incomplete table must not reject correct data.
    assertTrue(
        Validate.check(
                Map.of("anything", 1L), Schema.of("{\"k\":\"ref\",\"n\":\"Absent\"}"), Map.of())
            .isEmpty());
  }

  @Test
  void enforceThrowsInStrictModeAndIsSilentWhenOff() {
    Schema schema = Schema.of("{\"k\":\"obj\",\"f\":[[\"id\",{\"k\":\"str\"},1]]}");
    Object value = Map.of("id", 1L);
    Validate.enforce(value, schema, Map.of(), "widgets.get", ValidationMode.OFF);

    ResponseValidationException error =
        assertThrows(
            ResponseValidationException.class,
            () -> Validate.enforce(value, schema, Map.of(), "widgets.get", ValidationMode.STRICT));
    assertTrue(error.getMessage().contains("widgets.get"));
    assertEquals("widgets.get", error.operation());
  }

  @Test
  void aValidationFailureIsNotAnApiException() {
    // The server answered successfully; what failed is the contract between spec and
    // implementation. A caller
    // catching ApiException to handle "the API said no" must not swallow this.
    assertFalse(ApiException.class.isAssignableFrom(ResponseValidationException.class));
    assertTrue(SdkException.class.isAssignableFrom(ResponseValidationException.class));
  }
}

File: packages/runtime-php/tests/ValidateTest.php
<?php

declare(strict_types=1);

namespace Graft\Runtime\Tests;

use Graft\Runtime\ResponseValidationError;
use Graft\Runtime\Validate;
use Graft\Runtime\ValidationMode;
use PHPUnit\Framework\TestCase;

final class ValidateTest extends TestCase
{
    /**
     * @param array<string,mixed> $schema
     * @param array<string,mixed> $table
     * @return list<string>
     */
    private function check(mixed $value, array $schema, array $table = []): array
    {
        return Validate::check($value, $schema, $table);
    }

    public function testAcceptsMatchingPrimitives(): void
    {
        self::assertSame([], $this->check('x', ['k' => 'str']));
        self::assertSame([], $this->check(1, ['k' => 'int']));
        self::assertSame([], $this->check(1.5, ['k' => 'num']));
        self::assertSame([], $this->check(true, ['k' => 'bool']));
        self::assertSame([], $this->check(null, ['k' => 'null', 'i' => ['k' => 'str']]));
    }

    public function testNamesTheFieldAndBothTypes(): void
    {
        $problems = $this->check(
            ['id' => 42],
            ['k' => 'obj', 'f' => [['id', ['k' => 'str'], 1]]],
        );
        self::assertSame(['id should be a string but was an integer'], $problems);
    }

    public function testAcceptsAWholeFloatWhereAnIntegerIsDeclared(): void
    {
        // A JSON integer arrives as 1.0 from serializers with no integer type. Rejecting that would fail
        // on data that is actually correct.
        self::assertSame([], $this->check(1.0, ['k' => 'int']));
        self::assertNotSame([], $this->check(1.5, ['k' => 'int']));
    }

    public function testReportsAMissingRequiredFieldButNotAMissingOptionalOne(): void
    {
        $schema = ['k' => 'obj', 'f' => [['id', ['k' => 'str'], 1], ['name', ['k' => 'str']]]];
        self::assertSame(['id is missing'], $this->check(['name' => 'x'], $schema));
        self::assertSame([], $this->check(['id' => 'x'], $schema));
    }

    public function testNeverComplainsAboutUnknownFields(): void
    {
        // A server adding a field must not break a client. That is the whole point of an evolving API.
        self::assertSame([], $this->check(
            ['id' => 'x', 'brandNew' => 123],
            ['k' => 'obj', 'f' => [['id', ['k' => 'str'], 1]]],
        ));
    }

    public function testNeverChecksEnumMembership(): void
    {
        // Servers add enum values without warning; the open-enum rule exists so that is not a decode
        // failure. An enum is validated as its base type only.
        self::assertSame([], $this->check('a-value-added-later', ['k' => 'str']));
    }

    public function testTreatsAnEmptyArrayAsAValidEmptyMap(): void
    {
        // The PHP empty-map artifact, from the emitting side this time: a PHP backend serialises `{}` as
        // `[]`, and that is a valid empty map rather than a wrong type.
        self::assertSame([], $this->check([], ['k' => 'map', 'v' => ['k' => 'str']]));
    }

    public function testWalksArraysAndReportsTheIndex(): void
    {
        $problems = $this->check(['a', 2], ['k' => 'arr', 'i' => ['k' => 'str']]);
        self::assertSame(['[1] should be a string but was an integer'], $problems);
    }

    public function testAcceptsAnyBranchOfAUnionAndReportsOnceWhenNoneMatch(): void
    {
        $schema = ['k' => 'or', 'o' => [['k' => 'str'], ['k' => 'int']]];
        self::assertSame([], $this->check('x', $schema));
        self::assertSame([], $this->check(3, $schema));
        // One message, not one per branch: a union of five reporting five problems buries the real one.
        self::assertCount(1, $this->check(true, $schema));
    }

    public function testTerminatesOnASelfReferentialSchema(): void
    {
        // The cycle closes through the table rather than through recursion, so this is finite by
        // construction rather than by a depth cap.
        $table = ['Node' => ['k' => 'obj', 'f' => [['child', ['k' => 'ref', 'n' => 'Node']]]]];
        self::assertSame([], $this->check(
            ['child' => ['child' => ['child' => []]]],
            ['k' => 'ref', 'n' => 'Node'],
            $table,
        ));
    }

    public function testTreatsAMissingTableEntryAsAny(): void
    {
        // An incomplete table must not reject correct data.
        self::assertSame([], $this->check(['anything' => 1], ['k' => 'ref', 'n' => 'Absent']));
    }

    public function testEnforceThrowsInStrictModeAndIsSilentWhenOff(): void
    {
        $schema = ['k' => 'obj', 'f' => [['id', ['k' => 'str'], 1]]];
        Validate::enforce(['id' => 1], $schema, [], 'widgets.get', ValidationMode::Off);

        $this->expectException(ResponseValidationError::class);
        $this->expectExceptionMessageMatches('/widgets\.get/');
        Validate::enforce(['id' => 1], $schema, [], 'widgets.get', ValidationMode::Strict);
    }

    public function testAValidationErrorIsNotAnApiError(): void
    {
        // The server answered successfully; what failed is the contract between spec and implementation.
        // A caller catching ApiError to handle "the API said no" must not swallow this.
        // Read off the parent chain rather than asserted with instanceof or is_subclass_of: PHPStan
        // constant-folds both of those to always-true/always-false, which is the typechecker confirming
        // the hierarchy — but leaves nothing guarding a future change to it. `class_parents` it cannot
        // fold, so this stays a real test of the decision rather than a tautology.
        $parents = class_parents(ResponseValidationError::class);
        self::assertIsArray($parents);
        self::assertNotContains(\Graft\Runtime\ApiError::class, $parents);
        self::assertContains(\Graft\Runtime\SdkError::class, $parents);
    }
}
